File: examples/golang/src/example_with_zone/example_with_zone.go
// Copyright 2016, Cossack Labs Limited
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package main

import (
	"database/sql"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net/http"
	"time"

	"github.com/cossacklabs/acra/acra-writer"
	"github.com/cossacklabs/themis/gothemis/keys"
	_ "github.com/go-sql-driver/mysql"
	_ "github.com/lib/pq"
	log "github.com/sirupsen/logrus"
)

// AcraConnectorAddress default address
var AcraConnectorAddress = "http://127.0.0.1:9191"

// ZoneData stores zone: zoneID and PublicKey
type ZoneData struct {
	ID        string
	PublicKey []byte
}

func main() {
	mysql := flag.Bool("mysql", false, "Use MySQL driver")
	_ = flag.Bool("postgresql", false, "Use PostgreSQL driver (default if nothing else set)")
	dbname := flag.String("db_name", "acra", "Database name")
	host := flag.String("host", "127.0.0.1", "Database host")
	port := flag.Int("port", 9494, "Database port")
	user := flag.String("db_user", "test", "Database user")
	password := flag.String("db_password", "password", "Database user's password")
	data := flag.String("data", "", "Data to save")
	printData := flag.Bool("print", false, "Print data from database")
	zoneID := flag.String("zone_id", "", "Zone ID to fetch")
	flag.Parse()

	connectionString := fmt.Sprintf("user=%v password=%v dbname=%v host=%v port=%v", *user, *password, *dbname, *host, *port)
	driver := "postgres"
	if *mysql {
		// username:password@protocol(address)/dbname?param=value
		// https://github.com/go-sql-driver/mysql#dsn-data-source-name
		connectionString = fmt.Sprintf("%v:%v@tcp(%v:%v)/%v", *user, *password, *host, *port, *dbname)
		driver = "mysql"
	}

	db, err := sql.Open(driver, connectionString)
	if err != nil {
		log.Fatal(err)
	}

	err = db.Ping()
	if err != nil {
		log.Fatal(err)
	}
	if *mysql {
		query := "CREATE TABLE IF NOT EXISTS test_example_with_zone(id INTEGER PRIMARY KEY, zone BINARY(24), data VARBINARY(1000), raw_data VARCHAR(1000));"
		fmt.Printf("Create test_example_with_zone table with command: '%v'\n", query)
		_, err = db.Exec(query)
	} else {
		query := "CREATE TABLE IF NOT EXISTS test_example_with_zone(id INTEGER PRIMARY KEY, zone BYTEA, data BYTEA, raw_data TEXT);"
		fmt.Printf("Create test_example_with_zone table with command: '%v'\n", query)
		_, err = db.Exec(query)
	}
	if err != nil {
		panic(err)
	}

	if *data != "" {
		rand.Seed(time.Now().UnixNano())
		//get new zone over http
		resp, err := http.Get(fmt.Sprintf("%s/getNewZone", AcraConnectorAddress))
		if err != nil {
			panic("Error: getting new zone data (need start Acra with zones (-z) )")
		}
		defer resp.Body.Close()
		body, err := ioutil.ReadAll(resp.Body)
		if err != nil {
			panic(err)
		}
		zoneData := []byte(body)
		var parsedZoneData ZoneData
		err = json.Unmarshal(zoneData, &parsedZoneData)
		if err != nil {
			panic(err)
		}
		zonePublic := parsedZoneData.PublicKey
		zoneID := []byte(parsedZoneData.ID)

		acrastruct, err := acrawriter.CreateAcrastruct([]byte(*data), &keys.PublicKey{Value: zonePublic}, zoneID)
		if err != nil {
			panic(err)
		}
		fmt.Printf("saved with zone: %v\n", string(zoneID))
		if *mysql {
			_, err = db.Exec("insert into test_example_with_zone (id, zone, data, raw_data) values (?, ?, ?, ?);", rand.Int31(), zoneID, acrastruct, *data)
		} else {
			_, err = db.Exec("insert into test_example_with_zone (id, zone, data, raw_data) values ($1, $2, $3, $4);", rand.Int31(), zoneID, acrastruct, *data)
		}
		if err != nil {
			panic(err)
		}
	} else if *printData {
		var query string
		if *mysql {
			query = `SELECT ?, data, raw_data, zone FROM test_example_with_zone;`
		} else {
			query = `SELECT $1::bytea, data, raw_data, zone FROM test_example_with_zone;`
		}

		fmt.Printf("Select from db with command: '%v'\n", query)
		rows, err := db.Query(query, []byte(*zoneID))
		defer rows.Close()
		if err != nil {
			log.Fatal(err)
		}
		var zone, rawZone, data []byte
		var rawData string
		fmt.Println("zone, data - raw_data")
		for rows.Next() {
			err := rows.Scan(&zone, &data, &rawData, &rawZone)
			if err != nil {
				panic(err)
			}
			fmt.Printf("zone: %v\ndata: %v\nraw_data: %v\nrow zone: %v\n\n", string(zone), string(data), string(rawData), string(rawZone))
		}
	}
	fmt.Println("Finish")
}

File: examples/python/example_with_zone.py
# Copyright 2016, Cossack Labs Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# coding: utf-8
import argparse
from sqlalchemy import (Table, Column, Integer, MetaData, select, LargeBinary, Text, cast)
from sqlalchemy.dialects.postgresql import BYTEA
from acrawriter import create_acrastruct
from common import get_engine, get_default, get_zone, register_common_cli_params


def print_data(zone_id, connection):
    """fetch data from database (use zone_id if not empty/None) and print to
    console"""
    result = connection.execute(
        # explicitly pass zone id before related data
        select([cast(zone_id.encode('utf-8'), BYTEA), test_table]))
    result = result.fetchall()
    ZONE_ID_INDEX = 0
    print("use zone_id: ", zone_id)
    print("{:<3} - {} - {} - {}".format("id", 'zone', "data", "raw_data"))
    for row in result:
        print(
            "{:<3} - {} - {} - {}\n".format(
            row['id'], row[ZONE_ID_INDEX].decode('utf-8'),
            row['data'].decode('utf-8', errors='ignore'), row['raw_data']))


def write_data(data, connection):
    zone_id, key = get_zone()
    print("data: {}\nzone: {}".format(data, zone_id))

    # here we encrypt our data and wrap into AcraStruct
    encrypted_data = create_acrastruct(
        data.encode('utf-8'), key, zone_id.encode('utf-8'))

    connection.execute(
        test_table.insert(), data=encrypted_data,
        zone_id=zone_id.encode('utf-8'),
        raw_data=data)


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    register_common_cli_params(parser)
    parser.add_argument('--zone_id', type=str,
                        default=get_default('zone_id', ''),
                        help='Zone id for fetching data. Use only with --print '
                             'option')
    parser.add_argument('--data', type=str,
                        default=get_default('data', ''),
                        help='data to save in ascii. default random data')
    args = parser.parse_args()

    metadata = MetaData()
    test_table = Table(
        'test_example_with_zone', metadata,
        Column('id', Integer, primary_key=True, nullable=False),
        Column('zone_id', LargeBinary, nullable=True),
        Column('data', LargeBinary, nullable=False),
        Column('raw_data', Text, nullable=False),
    )

    engine = get_engine(
        db_host=args.host, db_port=args.port, db_user=args.db_user, db_password=args.db_password,
        db_name=args.db_name, is_mysql=args.mysql, is_postgresql=args.postgresql,
        tls_ca=args.tls_root_cert, tls_key=args.tls_key, tls_crt=args.tls_cert,
        sslmode=args.ssl_mode, verbose=args.verbose)
    connection = engine.connect()
    metadata.create_all(engine)

    if args.print:
        print_data(args.zone_id, connection)
    elif args.data:
        if args.zone_id:
            print("To encrypt data script will generate new zone and print "
                  "zone id with public key after execution. Don't use "
                  "--zone_id option with --data option.")
            exit(1)
        write_data(args.data, connection)
    else:
        print('Use --print or --data options')
        exit(1)
